test/extensions/filters/udp/dns_filter/dns_filter_test_utils.cc

```diff
@@ -84,6 +88,346 @@ void verifyAddress(const std::list<std::string>& addresses, const DnsAnswerRecor
   ASSERT_TRUE(iter != addresses.end());
 }
 
+absl::string_view getProtoFromName(const absl::string_view name) {
+  size_t start = name.find_first_of('.');
+  if (start != std::string::npos && ++start < name.size() - 1) {
+    if (name[start] == '_') {
+      const size_t offset = name.find_first_of('.', ++start);
+      if (start != std::string::npos && offset < name.size()) {
+        return name.substr(start, offset - start);
+      }
+    }
+  }
+  return EMPTY_STRING;
+}
+
+DnsQueryContextPtr DnsResponseValidator::createResponseContext(Network::UdpRecvData& client_request,
```

Comment:
Is this function is added for test only? And is it to test some testing code, not production code?
I am thinking we should just test createQueryContext().

Reply:
This function is to parse the DNS response message, and it is used to verify the response generated by DNS filter, which we want to test. Basically, the query and response are sharing the same message format, this is only to parse and verify the response.

Reply:
I understand what you mean now.  So, the test is to send a request, get the response and parse the response and verify it is expected. The original code is using createQueryContext() to parse the response since that logic is embedded there.  With the new code, createResponseContext is added to specially handle the response.  Also, in the real world, Envoy won't parse the response. Only the DNS client who sends the request will parse the response. Thus createResponseContext() is only called in Envoy test code.

With that, I think the only thing I recommend is to add a few new test cases to verify the new logic --- bail out early and return false if answer is detected in query, which you added in createQueryContext()->parseDnsObject().

Reply:
More generally, I'm wondering if there is some way to actually replace this custom code with c-ares? I bet there is some way we could pipe the response into c-ares and have it parse it. IMO That would be better than maintaining this code just for testing purposes. Maybe add a TODO around this also?

Reply:
Good point, DNS filter only supports A, AAAA, SRV records, so we could use c-ares methods like [ares_parse_a_reply](https://c-ares.org/ares_parse_a_reply.html), [ares_parse_aaaa_reply](https://c-ares.org/ares_parse_aaaa_reply.html), and [ares_parse_srv_reply](https://c-ares.org/ares_parse_srv_reply.html). 

This is only for validation, IMO it is okay to keep as it is, but will add a TODO.